### DISCUSSION TITLE
How to know if a KtExpression can change it's

### DISCUSSION THREAD

**Discussion by BraisGabin** [detekt/detekt#4291] [Q&A]
There are some `KtExpression`s that the compiler knows that if you evaluate them again they will not change. Examples are: local `val`s, local `var`s (if that var is not used in any previous lambda), val properties without a custom `get`, etc. There are other cases, for example a funcation call that the compiler can't know. How could I get that information?

I need it for two different rules: `UnnecessaryLet` and to implement the rule defined in #3353.

If you don't know the exact answer but you know where I could find examples of this in our code (I think that we aren't using this yet) or in the kotlin compiler code would be great. I know that the kotlin compiler knows this because it needs it for the smart cast.

**Comment by cortinico** [comment]
> There are other cases, for example a funcation call that the compiler can't know. How could I get that information?

I'm unsure I got your questions. What is going to change? Are you referring to `KtExpression` and mutable variables?

**Reply by BraisGabin** [reply]
I have a `KtExpression` and I want to know if I evaluate that expression multiple times the reference that I get can be different between invocations. Examples:

```kotlin
fun asdf() {
  var a = someghing()
  a?.let { /* ... */ } // In this case a can't change it's reference, even it is a `var`.
}
```
```kotlin
fun asdf() {
  var a = someghing()
  GlobalScope.launch { a = something() }
  a?.let { /* ... */ } // In this case a can change it's reference. Because it is unsed in a lambda.
}
```
```kotlin
class A(var a: String?) {
  fun asdf() {
    a?.let { /* ... */ } // In this case a can change it's reference. Another thread could do it
  }
}
```
```kotlin
class A(val a: String?) {
  fun asdf() {
    a?.let { /* ... */ } // In this case a can not change it's reference. It's val
  }
}
```

For me it's important if the reference can change or not because, based on that, I can decide if the usage of `let` is unnecesary.
